pkg/threescale/authorizer/authorizer.go

```diff
@@ -0,0 +1,276 @@
+	client, err := m.clientBuilder.BuildBackendClient(backendURL)
+	if err != nil {
+		return nil, fmt.Errorf("unable to build required client for 3scale backend - %s", err.Error())
+	}
+
+	req, err := request.ToAPIRequest()
+	if err != nil {
+		return nil, fmt.Errorf("unable to build request to 3scale - %s", err)
+	}
+
+	res, err := client.AuthRep(*req)
+	if err != nil {
+		return nil, fmt.Errorf("error calling AuthRep - %s", err)
+	}
+
+	return &BackendResponse{Authorized: res.Success()}, nil
+}
+
+// validateSystemRequest to avoid wasting compute time on invalid request
+func (m Manager) validateSystemRequest(request SystemRequest) error {
+	if request.Environment == "" || request.ServiceID == "" || request.AccessToken == "" {
+		return fmt.Errorf("invalid arguements provided")
+	}
+	return nil
+}
+
+func (m Manager) generateSystemCacheKey(systemURL, svcID string) string {
+	return fmt.Sprintf("%s_%s", systemURL, svcID)
+}
+
+func (m Manager) fetchSystemConfigFromCache(systemURL string, request SystemRequest) (client.ProxyConfig, error) {
+	var config client.ProxyConfig
+	var err error
+
+	cacheKey := m.generateSystemCacheKey(systemURL, request.ServiceID)
+	cachedValue, found := m.systemCache.cache.Get(cacheKey)
+	if !found {
+		config, err = m.fetchSystemConfigRemotely(systemURL, request)
+		if err != nil {
+			return config, err
+		}
+
+		itemToCache := &cache.Value{Item: config}
+		itemToCache = m.setValueFromConfig(systemURL, request, itemToCache)
+		m.systemCache.cache.Set(cacheKey, *itemToCache)
+
+	} else {
+		config = cachedValue.Item
+	}
+
+	return config, err
+}
+
+func (m Manager) fetchSystemConfigRemotely(systemURL string, request SystemRequest) (client.ProxyConfig, error) {
+	var config client.ProxyConfig
+
+	systemClient, err := m.clientBuilder.BuildSystemClient(systemURL, request.AccessToken)
+	if err != nil {
+		return config, fmt.Errorf("unable to build system client for %s - %s", systemURL, err.Error())
+	}
+
+	proxyConfElement, err := systemClient.GetLatestProxyConfig(request.ServiceID, request.Environment)
+	if err != nil {
+		return config, fmt.Errorf("unable to fetch required data from 3scale system - %s", err.Error())
+	}
+
+	return proxyConfElement.ProxyConfig, nil
+}
+
+func (m Manager) refreshCallback(systemURL string, request SystemRequest, retryAttempts int) func() (client.ProxyConfig, error) {
+	return func() (client.ProxyConfig, error) {
+		config, err := m.fetchSystemConfigRemotely(systemURL, request)
+		if err != nil {
+			if retryAttempts > 0 {
+				retryAttempts--
+				return m.refreshCallback(systemURL, request, retryAttempts)()
+			}
+		}
+		return config, err
+	}
+}
+
+func (m Manager) setValueFromConfig(systemURL string, request SystemRequest, value *cache.Value) *cache.Value {
+	value.SetRefreshCallback(m.refreshCallback(systemURL, request, m.systemCache.config.numRetryFailedRefresh))
+	value.SetExpiry(time.Now().Add(time.Second * m.systemCache.config.ttlSeconds))
```

Comment:
I am trying to figure out here if this `m.systemCache.config.ttlSeconds` is any different than the initial value we use to build the cache. Does that ever change? If not it would seem that this ttl is just the same ttl stored in the cache.

Reply:
yes thats correct really, i just added it here to make the behaviour explicit but I can remove that if you think its best?

Reply:
Yes, it confused me quite a bit, because my natural reaction was: we keep two ttls because they vary and somehow we must keep the original. Same thing for any other things we might be duplicating without a need because they don't vary.

Reply:
gotcha